java/readme.md

```diff
@@ -0,0 +1,224 @@
+    }
+
+    public static void main(String[] args) throws InterruptedException, ExecutionException {
+        createTopic("demo-topic");
+    }
+
+    static void createTopic(String topicName) throws InterruptedException, ExecutionException {
+        try (AdminClient adminClient = AdminClient.create(clientProperties())) {
+            adminClient.createTopics(Collections.singleton(new NewTopic(topicName, Optional.of(1), Optional.empty())))
+                    .all()
+                    .get();
+        } catch (ExecutionException | InterruptedException e) {
+            if (!(e.getCause() instanceof TopicExistsException)) {
+                throw e;
+            }
+            System.out.println("topic already exists");
+        }
+    }
+}
+```
+
+
+## Create a producer to send messages
+Create a file named `src/main/java/org/example/Producer.java` and paste the code below.
+
+```java title="Main.java"
+package org.example;
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.concurrent.ExecutionException;
+
+public class Producer {
+    public static void main(String[] args) {
+        produce("demo-topic");
+    }
+
+    public static void produce(String topicName) {
+        try (org.apache.kafka.clients.producer.Producer<Void, String> producer = new KafkaProducer<>(Main.clientProperties())) {
+            for (var i = 0; i < 100; i++) {
+                ProducerRecord<Void, String> record = new ProducerRecord<>(topicName, String.format("asynchronous message #%d", i));
+                producer.send(record, (r, e) -> {
+                    if (e == null) {
+                        System.out.printf("Sent to topic '%s' at offset %d\n", r.topic(), r.offset());
+                    } else {
+                        System.out.println("Error sending message: " + e);
+                    }
+                });
+
+            }
+            producer.flush();
+        }
+    }
+}
+```
+
+
+## Create a consumer to read data from the topic
+Create a file named `src/main/java/org/example/Consumer.java` and paste the code below. Note that the consumer requires slightly more configuration than the producer.
+
+```java title="Main.java"
+package org.example;
+
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+
+import java.time.Duration;
+import java.util.*;
+
+public class Consumer {
+    public static void main(String[] args) {
+        consume("demo-topic");
+    }
+
+    public static void consume(String topicName) {
+        var props = Main.clientProperties();
+        var consumerGroup = "basic-group-" + UUID.randomUUID();
+        // Additional configuration
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroup);
```

Comment:
I'm not 100% sure we can - the client complained about not having a group configured. However, I did hardcode the name in the end (to "demo-group") because we need some ACL instructions for that too.

Reply:
Turns out we can - setting enable.auto.commit=false is the ticket here (it seems to default to true)